### PR TITLE
Dedupe / Refactor `.emails` on OSFUser [OSF-6072]

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
Enforce email uniqueness at the database level

## Changes
* Add email model, points to users
* Remove explicit `emails` field
* Migrate
* Update syntax accordingly

## Side effects
Migration requires an estimated 10 minutes of downtime

## Ticket
[[OSF-6072]](https://openscience.atlassian.net/browse/OSF-6072)

### CODE REVIEW

**Comment by sloria** [comment]
Just wondering; how long did the migration take to complete in your local run?

File: admin/templates/users/user.html [hunk 1]
```diff
@@ -139,7 +139,7 @@
                 <tr>
                     <td>Email</td>
                     <td>
-                        {% for email in user.emails %}
+                        {% for email in user.emails.values_list('address', flat=True) %}
                             <li>
                                 {% if forloop.first %}
```
Comment (mfraezz) [inline admin/templates/users/user.html]: It held during my testing, but may not always.

Also, yes.

File: osf/models/user.py [hunk 2]
```diff
@@ -102,6 +102,14 @@ def create_superuser(self, username, password):
         return user
 
 
+class Email(BaseModel, TimeStampedModel):
+    address = LowercaseEmailField(unique=True, db_index=True, validators=[validate_email])
+    user = models.ForeignKey('OSFUser', related_name='emails', null=True, blank=True, on_delete=models.CASCADE)
```
Comment (mfraezz) [inline osf/models/user.py]: Currently, it should not. The PR this was originally inspired by had email objects associated with models other than just`OSFUser`. I can remove nullability until that is true again

File: osf/utils/fields.py [hunk 3]
```diff
@@ -21,6 +21,17 @@ def get_prep_value(self, value):
             value = value.lower()
         return value
 
+class LowercaseEmailField(models.EmailField):
+    # TODO: This is technically not compliant with RFC 822, which requires
```
Comment (mfraezz) [inline osf/utils/fields.py]: Fair.

This text exists in at least one other place in the codebase, should it be changed there too?

File: osf/utils/fields.py [hunk 4]
```diff
@@ -21,6 +21,17 @@ def get_prep_value(self, value):
...
+    #       that case be preserved in the "local-part" of an address. From
+    #       a practical standpoint, the vast majority of email servers do
+    #       not preserve case.
+    #       ref: https://tools.ietf.org/html/rfc822#section-6
+    def get_prep_value(self, value):
+        value = super(models.EmailField, self).get_prep_value(value)
+        if value is not None:
+            value = value.lower()
```
Comment (mfraezz) [inline osf/utils/fields.py]: Probably

File: osf/utils/fields.py (hunk 3 above)
Comment (sloria) [inline osf/utils/fields.py]: Yes

**Comment by mfraezz** [comment]
> Just wondering; how long did the migration take to complete in your local run?

About 15 minutes on my horribly slow machine, so probably less than 5 on anything decent. I'll speed it up with `values_list` for the `OSFUser` select.

**Comment by mfraezz** [comment]
Merge build failed due to upstream errors, [personal TCI](https://travis-ci.org/mfraezz/osf.io/builds/233350531) passed.

**Comment by mfraezz** [comment]
Build failed due only to upstream errors, merging

**Pull request merged**